### PR TITLE
Rollover CFD's automatically when a configurable rollover offset has …

### PR DESCRIPTION
**Pull request by rishflab**
…elapsed

The rollover offset is the number of settlement intervals after which rollover is triggered.

### CODE REVIEW

File: daemon/src/model/cfd.rs [hunk 1]
```diff
@@ -680,20 +680,26 @@ impl Cfd {
         self.order.oracle_event_id.timestamp()
     }
 
-    /// A factor to be added to the CFD order settlement_time_interval_hours for calculating the
+    pub fn creation_timestamp(&self) -> Result<OffsetDateTime> {
+        let timestamp =
+            OffsetDateTime::from_unix_timestamp(self.order.creation_timestamp.seconds())?;
+        Ok(timestamp)
+    }
```
Comment (thomaseizinger) [inline daemon/src/model/cfd.rs]: I think this might be dead code now?

File: daemon/src/lib.rs [hunk 2]
```diff
@@ -254,6 +256,12 @@ where
 
         tokio::spawn(oracle_ctx.run(oracle_constructor(cfds, Box::new(fan_out_actor))));
 
+        // TODO: Optimize initial auto-rollover to take time into account so that we don't want to
+        // trigger auto-rollover after an immediate restart
+        cfd_actor_addr
```
Comment (bonomat) [inline daemon/src/lib.rs]: Why should we not trigger it immediately after start?

File: daemon/src/taker_cfd.rs [hunk 3]
```diff
@@ -60,6 +60,21 @@ enum RollOverState {
     None,
 }
 
+pub struct RollOverParams {
+    settlement_interval: Duration,
+    /// The time since cfd creation where rollover is triggered
```
Comment (bonomat) [inline daemon/src/taker_cfd.rs]: I think this doc is outdated.

File: daemon/src/taker_cfd.rs [hunk 4]
```diff
@@ -167,6 +185,65 @@ impl<O, M, W> Actor<O, M, W> {
...
+        for cfd in cfds
+            .iter()
+            .filter(|cfd| matches!(cfd.state, CfdState::Open { .. }))
+        {
+            let now = OffsetDateTime::now_utc();
+
+            let elapsed_since_creation =
+                now - (cfd.expiry_timestamp() - self.roll_over_params.settlement_interval);
+
+            // Two conditions need to be bet for a roll over to be triggered.
```
Comment (bonomat) [inline daemon/src/taker_cfd.rs]:
```suggestion
            // Two conditions need to be met for a roll over to be triggered.
```

File: daemon/src/taker_cfd.rs [hunk 5]
```diff
@@ -167,6 +185,65 @@ impl<O, M, W> Actor<O, M, W> {
...
+            // The cfd should not of run longer that the settlement interval.
```
Comment (bonomat) [inline daemon/src/taker_cfd.rs]:
```suggestion
            // The cfd should not run longer that the settlement interval.
```

File: daemon/src/taker_cfd.rs [hunk 6]
```diff
@@ -60,6 +60,21 @@ enum RollOverState {
     None,
 }
 
+pub struct RollOverParams {
```
Comment (thomaseizinger) [inline daemon/src/taker_cfd.rs]: I think these concern themselves with our "auto" roll-over feature, not roll-over per se.

cc @da-kami

```suggestion
pub struct AutoRollOverParams {
```

File: daemon/src/taker_cfd.rs [hunk 7]
```diff
@@ -60,6 +60,21 @@ enum RollOverState {
...
+    settlement_interval: Duration,
+    /// The time since cfd creation where rollover is triggered
+    rollover_interval: Duration,
```
Comment (thomaseizinger) [inline daemon/src/taker_cfd.rs]: To me, the docs and variable name here still don't make sense to me.

As far as I understand, this interval defines the minimum time that has to pass for a CFD to be automatically rolled over. How about this:

```suggestion
    /// Defines the minimum duration a CFD needs to be open for until a roll-over is triggered.
    min_open_until_rollover: Duration,
```

cc @da-kami @bonomat

File: daemon/src/taker_cfd.rs [hunk 8]
```diff
@@ -60,6 +60,21 @@ enum RollOverState {
     None,
 }
 
+pub struct RollOverParams {
+    settlement_interval: Duration,
```
Comment (thomaseizinger) [inline daemon/src/taker_cfd.rs:65]:
```suggestion
    /// Defines the maximum duration a CFD may be open for to still qualify for automatic roll-over.
    max_open: Duration,
```

cc @da-kami @bonomat

File: daemon/src/lib.rs [hunk 9]
```diff
@@ -85,7 +86,7 @@ where
             Box<dyn MessageChannel<NewTakerOnline>>,
             Box<dyn MessageChannel<FromTaker>>,
         ) -> T,
-        settlement_time_interval_hours: time::Duration,
+        settlement_interval: time::Duration,
```
Comment (thomaseizinger) [inline daemon/src/lib.rs]: I know the name is ugly but it would be great to minimize the diff of this PR to just the auto-roll over feature. Can you please undo this rename and squash all commits afterwards?

Feel free to send a follow-up PR that cleans up these names.

File: daemon/src/taker_cfd.rs [hunk 10]
```diff
@@ -167,6 +185,65 @@ impl<O, M, W> Actor<O, M, W> {
...
+        self.current_pending_proposals.clear();
+
+        // Only cfds in state `Open` are allowed for auto-rollover proposal
+        for cfd in cfds
+            .iter()
+            .filter(|cfd| matches!(cfd.state, CfdState::Open { .. }))
+        {
+            let now = OffsetDateTime::now_utc();
+
+            let elapsed_since_creation =
```
Comment (thomaseizinger) [inline daemon/src/taker_cfd.rs]: I think this variable name is now stale.

File: daemon/src/taker_cfd.rs [hunk 11]
```diff
@@ -167,6 +185,65 @@ impl<O, M, W> Actor<O, M, W> {
...
+        self.send_pending_update_proposals()?;
+
+        self.send_to_maker
+            .send(wire::TakerToMaker::ProposeRollOver {
+                order_id: proposal.order_id,
+                timestamp: proposal.timestamp,
+            })
+            .await?;
+
+        tracing::debug!("proposed roll over");
```
Comment (thomaseizinger) [inline daemon/src/taker_cfd.rs:242]: I think this is important enough to warrant an `info` log, plus it would be good to include some context.

```suggestion
        tracing::info!(%order_id, "Proposed roll over");
```

File: daemon/tests/harness/mocks/mod.rs [hunk 12]
```diff
@@ -73,12 +73,26 @@ impl Mocks {
...
+            .expect_monitor_attestation()
+            .return_const(());
+    }
+
+    pub async fn mock_start_monitoring(&mut self) {
+        self.monitor()
+            .await
+            .expect_start_monitoring()
+            .return_const(());
+    }
```
Comment (thomaseizinger) [inline daemon/tests/harness/mocks/mod.rs]: I think we should include these in `mock_common_empty_handlers` otherwise the test is super noisy.

File: daemon/tests/happy_path.rs [hunk 13]
```diff
@@ -59,6 +64,150 @@ async fn taker_takes_order_and_maker_rejects() {
...
+        .expect_sign()
+        .times(1)
+        .returning(|sign| Ok(sign.psbt))
+        .in_sequence(&mut seq);
+    wallet
+        .expect_broadcast()
+        .times(1)
+        .returning(|_| Ok(dummy_tx_id()))
+        .in_sequence(&mut seq);
+}
```
Comment (thomaseizinger) [inline daemon/tests/happy_path.rs]: This is stale, we have a function on `Mocks` that does the same thing.

File: daemon/tests/happy_path.rs [hunk 14]
```diff
@@ -59,6 +64,150 @@ async fn taker_takes_order_and_maker_rejects() {
...
+        .cfd_actor_addr
+        .send(monitor::Event::LockFinality(taker_cfd.order.id))
+        .await
+        .unwrap();
+
+    let (taker_cfd_open_state, maker_cfd_open_state) =
+        next_cfd(&mut taker.cfd_feed, &mut maker.cfd_feed).await;
+
+    assert!(matches!(taker_cfd_open_state.state, CfdState::Open { .. }));
+    assert!(matches!(maker_cfd_open_state.state, CfdState::Open { .. }));
```
Comment (thomaseizinger) [inline daemon/tests/happy_path.rs]: All of this is essentially setup code, at least from the perspective of this test. Please create a helper like `start_both_in_open` that does all of this minus the assertions (setup code should not contain assertions).

File: daemon/tests/harness/mod.rs [hunk 15]
```diff
@@ -134,6 +147,29 @@ impl Maker {
...
+            .send(CfdAction::AcceptRollOver { order_id: order.id })
+            .await
+            .unwrap()
+            .unwrap();
+    }
+
+    pub async fn get_update_cfd_proposal(
+        &mut self,
+        order_id: &OrderId,
+    ) -> anyhow::Result<UpdateCfdProposal> {
```
Comment (thomaseizinger) [inline daemon/tests/harness/mod.rs]: No need to qualify `Result`.

```suggestion
    ) -> Result<UpdateCfdProposal> {
```

File: daemon/src/lib.rs (hunk 9 above)
Comment (thomaseizinger) [reply daemon/src/lib.rs]: For future PRs, can you please quickly scan the entire diff of the PR yourself just before you request a review? :)
These things are easy to spot!

File: daemon/src/taker_cfd.rs [hunk 16]
```diff
@@ -167,6 +185,65 @@ impl<O, M, W> Actor<O, M, W> {
...
+            let now = OffsetDateTime::now_utc();
+
+            let elapsed_since_creation =
+                now - (cfd.expiry_timestamp() - self.roll_over_params.settlement_interval);
+
+            // Two conditions need to be bet for a roll over to be triggered.
+            // The cfd should not of run longer that the settlement interval.
+            // The cfd should of lived longer than the rollover interval.
+            if self.roll_over_params.settlement_interval > elapsed_since_creation
+                && elapsed_since_creation < self.roll_over_params.rollover_interval
```
Comment (thomaseizinger) [inline daemon/src/taker_cfd.rs]: You could simplify this by moving this check _onto_ the `RollOverParams` struct. Note how you are accessing self.roll_over_params` twice in here.

This is an instance of _Tell, don't ask_.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Thanks, some more comments!

File: daemon/src/taker_cfd.rs (hunk 8 above)
Comment (bonomat) [reply daemon/src/taker_cfd.rs:65]: I would prefer `settlement_interval` or something with `*interval` in it.

File: daemon/src/taker_cfd.rs (hunk 7 above)
Comment (bonomat) [reply daemon/src/taker_cfd.rs]: I find it confusing if I have to think about the expiry from the beginning. 
I would rather think about it from the end, i.e. the CFD should always have a remaining expiry of 24h, if it doesn't, roll over.

File: daemon/src/taker_cfd.rs [hunk 17]
```diff
@@ -167,6 +185,65 @@ impl<O, M, W> Actor<O, M, W> {
...
+        {
+            let now = OffsetDateTime::now_utc();
+
+            let elapsed_since_creation =
+                now - (cfd.expiry_timestamp() - self.roll_over_params.settlement_interval);
+
+            // Two conditions need to be bet for a roll over to be triggered.
+            // The cfd should not of run longer that the settlement interval.
+            // The cfd should of lived longer than the rollover interval.
+            if self.roll_over_params.settlement_interval > elapsed_since_creation
```
Comment (bonomat) [inline daemon/src/taker_cfd.rs]: I find it confusing if I have to think about the time elapsed since the beginning. 
.
I would rather think about it from the end, i.e. the CFD should always have a remaining expiry of 24h, if it doesn't, roll over.
e.g. 
```
if cfd.expiry_timestamp() - now() <= self.roll_over_params.settlement_interval
  rollover()
```

File: daemon/src/taker_cfd.rs (hunk 8 above)
Comment (thomaseizinger) [reply daemon/src/taker_cfd.rs:65]: Why interval though? An interval is a duration that happens regularly. This value is not an interval. If you look at the logic how this is used, it only represents a cut-off time after which we no longer trigger roll over.

File: daemon/src/taker_cfd.rs (hunk 7 above)
Comment (thomaseizinger) [reply daemon/src/taker_cfd.rs]: >  I would rather think about it from the end, i.e. the CFD should always have a remaining expiry of 24h, if it doesn't, roll over.

That is not how it is currently implemented though. I based my suggestion on how things are implemented. If we implement it differently, I am happy to go with a name such as `min_time_to_expiry`.

Comment (thomaseizinger) [reply daemon/src/taker_cfd.rs]: Do we even need this? Shouldn't this always be the same value as `Order#settlement_time_interval_hours`?

File: daemon/src/lib.rs (hunk 9 above)
Comment (rishflab) [reply daemon/src/lib.rs]: sorry thats my bad

File: daemon/src/taker_cfd.rs (hunk 10 above)
Comment (rishflab) [reply daemon/src/taker_cfd.rs]: do you mean that it should be renamed? how about `elapsed_since_open`

Comment (thomaseizinger) [reply daemon/src/taker_cfd.rs]: You are using `cfd.expiry_timestamp()` yeah? So this computation is no longer based on the _creation_ timestamp. Or am I missing something?

File: daemon/src/lib.rs [hunk 18]
```diff
@@ -254,6 +256,12 @@ where
 
         tokio::spawn(oracle_ctx.run(oracle_constructor(cfds, Box::new(fan_out_actor))));
 
+        // TODO: Optimize initial auto-rollover to take time into account so that we don't want to
```
Comment (da-kami) [inline daemon/src/lib.rs]: I think this is an old ToDo from when I worked on it that is outdated now that we handle the roll-over based on the time left until the CFD settles. (A previous version of the code did not implement that, that's when the the ToDo was written).

File: daemon/src/taker_cfd.rs (hunk 8 above)
Comment (da-kami) [reply daemon/src/taker_cfd.rs:65]: I think in the context of rollover  `max_open` does make more sense than `settlement_interval`. 

---

A few words about different name on the "same" thing in a different context. 

Similar to the oracle's `announcement_lookahead` this variable has to be aligned with the `settlement_interval`. 
Initially we had a `term` that was pulled through the codebase and it was clear what it was, because it was named the same everywhere. That caused confusion in the context of a specific actor, but was easy to understand in "the big picture". Now I feel we have the opposite - in the context of each actor we try to find a name, but the big picture is hard to grasp.

I am not sure we can call it "an implementation detail" that the cfd's `settlement_interval` has to be used for the oracle's `announcement_lookahead` and the rollover's `max_open` - if those values are not aligned weird side effects can happen because we use different time constraints that are supposed to be aligned.

That said, maybe we should take a step back and think about creating a component that encapsulates this knowledge. I don't want to block this PR on this, but this is not the first time we run into this issue.

File: daemon/src/taker_cfd.rs [hunk 19]
```diff
@@ -60,6 +60,29 @@ enum RollOverState {
...
+
+impl AutoRolloverParams {
+    pub fn new(rollover_interval: Duration, settlement_interval: Duration) -> Result<Self> {
+        Ok(Self {
+            settlement_interval,
+            max_open_until_rollover: rollover_interval,
+        })
+    }
+
+    // Two conditions need to be bet for a roll over to be triggered.
```
Comment (da-kami) [inline daemon/src/taker_cfd.rs]:
```suggestion
    // Two conditions need to be met for a roll over to be triggered.
```

or can one use `bet` in this context in English?

File: daemon/src/taker_cfd.rs [hunk 20]
```diff
@@ -60,6 +60,29 @@ enum RollOverState {
...
+    // The cfd should not of run longer that the settlement interval.
+    // The cfd should of lived longer than the rollover interval.
+    pub fn is_cfd_due_for_rollover(&self, elapsed_since_cfd_opened: Duration) -> bool {
```
Comment (da-kami) [inline daemon/src/taker_cfd.rs]: I would find it more helpful if the complete time handling would be encapsulated in this function. 
i.e. I would pass in `cfd.expiry_timestamp()` as parameter into the function and all the rest should happen inside this function.

I find the current split up of logic is very hard to follow. Maybe that's a personal thing

File: daemon/tests/happy_path.rs (hunk 14 above)
Comment (rishflab) [reply daemon/tests/happy_path.rs]: I am not sure how to do this. in order to have setup and assertions seperate I would have to return multiple states from the setup function and assert on them inside the test. I think this might be a bit confusing.

File: daemon/src/taker_cfd.rs [hunk 21]
```diff
@@ -60,6 +60,29 @@ enum RollOverState {
...
+        self.settlement_interval > elapsed_since_cfd_opened
```
Comment (da-kami) [inline daemon/src/taker_cfd.rs]: Just to make sure I understand correct.

The `settlement_interval` is a constant set to `24` hours. This allows rollover if that constant is greater than the time elapsed since the CFD was opened. 
So, if `elapsed_since_cfd_opened` is `25` hours we won't roll over.

File: daemon/src/taker_cfd.rs [hunk 22]
```diff
@@ -60,6 +60,29 @@ enum RollOverState {
...
+impl AutoRolloverParams {
+    pub fn new(rollover_interval: Duration, settlement_interval: Duration) -> Result<Self> {
+        Ok(Self {
+            settlement_interval,
+            max_open_until_rollover: rollover_interval,
+        })
+    }
+
+    // Two conditions need to be bet for a roll over to be triggered.
+    // The cfd should not of run longer that the settlement interval.
```
Comment (da-kami) [inline daemon/src/taker_cfd.rs]:
```suggestion
    // The cfd should not run longer that the settlement interval.
```

File: daemon/src/taker_cfd.rs [hunk 23]
```diff
@@ -60,6 +60,29 @@ enum RollOverState {
...
+    // The cfd should of lived longer than the rollover interval.
```
Comment (da-kami) [inline daemon/src/taker_cfd.rs]: ??? - this sentence is somehow incomplete and does not reflect the (re-)names of variables anymore I feel.

File: daemon/src/taker_cfd.rs [hunk 24]
```diff
@@ -60,6 +60,29 @@ enum RollOverState {
     None,
 }
 
+pub struct AutoRolloverParams {
+    settlement_interval: Duration,
+    /// Defines the minimum duration a CFD needs to be open for until a roll-over is triggered.
+    max_open_until_rollover: Duration,
```
Comment (da-kami) [inline daemon/src/taker_cfd.rs]: I find it confusing that the comment states `minimum time ...` but then the variable is called `max_...` 

This represents the `1 hour` a CFD has to be at least open until we trigger a rollover. Maybe just remove the `max` ? 

Sry, I think you had a discussion with @thomaseizinger on that one before. If you decide to keep the variable name as is it would still be nice to align the comment.

**Review by da-kami — CHANGES_REQUESTED** [review]
My initial work included removing the configurable settlement interval from the maker.

See: https://github.com/itchysats/itchysats/commit/9e9ed8632db2674206804ff6cfb34baa387efb83#L55

Was it a conscious decision to not include this into this PR, or was it just accidentally dropped?

We don't actively use this feature and with introducing logic into the taker, that is unaware of what the maker set, this can cause weird side effects.
I think we should remove the configurable settlement interval as part of this PR and use **one** constant that represents what is called `settlement_interval` for both the maker and taker in the production code.

File: daemon/tests/happy_path.rs (hunk 14 above)
Comment (thomaseizinger) [reply daemon/tests/happy_path.rs]: I think it is fine to have setup functions that don't have any assertions. We have a separate test that has detailed assertions things that are meant to happen on the way to an open CFD.

We can have a setup function that assumes that this works and just creates an open CFD. If anything breaks within that part of the code base, all these tests will fail (and probably with different errors). It is then up to the dev to check which test is the one that points out the problem with the most detail and which tests are just a symptom. That should be easy. If something panics in the setup, no need to investigate that test.

File: daemon/src/taker_cfd.rs (hunk 8 above)
Comment (thomaseizinger) [reply daemon/src/taker_cfd.rs:65]: I don't think it has to be aligned, has it? `max_open` should be **equal or less** than `settlement_interval`. 

> That caused confusion in the context of a specific actor,

I don't think I was part of this conversation. Can you provide more context?

If a value MUST be the same across multiple parts of the system for it to work then we should absolutely have it as a single value.

File: daemon/src/taker_cfd.rs (hunk 24 above)
Comment (thomaseizinger) [reply daemon/src/taker_cfd.rs]: My original suggestion called it `min_open_until_rollover` :D

File: daemon/src/taker_cfd.rs (hunk 17 above)
Comment (da-kami) [reply daemon/src/taker_cfd.rs]: For me this is also easier to follow.

File: daemon/src/lib.rs (hunk 2 above)
Comment (rishflab) [reply daemon/src/lib.rs]: I think this is stale now as we check if the `time_until_rollover` has elapsed before triggering rollover. Can you confirm @da-kami

File: daemon/tests/happy_path.rs (hunk 14 above)
Comment (rishflab) [reply daemon/tests/happy_path.rs]: out of curiosity what is the problem in leaving the assertions in the setup functions. Wont it make it easier to debug in the case something goes wrong in the future?

File: daemon/src/taker_cfd.rs (hunk 17 above)
Comment (rishflab) [reply daemon/src/taker_cfd.rs]: ok i will flip it around

File: daemon/src/taker_cfd.rs (hunk 8 above)
Comment (da-kami) [reply daemon/src/taker_cfd.rs:65]: > I don't think it has to be aligned, has it? max_open should be equal or less than settlement_interval.

It has to be equal to make sense. It *can* be less for the system to still work, but that is out of the context of what we want to achieve. 

> I don't think I was part of this conversation. Can you provide more context?

Yes, this happened when you were on vacation 😬
Here is the PR:
https://github.com/itchysats/itchysats/pull/424

> If a value MUST be the same across multiple parts of the system for it to work then we should absolutely have it as a single value.

I am not sure the rename was a good idea. The rename is correct in the context of each actor, but the fact that the same value has to be used gets lost. Theoretically it does not have to be *exactly the same* as you pointed out above, but in the context of this system it *should* be. 

@rishflab and me are adding a commit that uses exactly *one* constant to represent the value (i.e. the 24 hours).
I hope that helps. My initial implementation of auto-rollover included that, but somehow it was lost on the way. I should have structured the commits more explicitly.

File: daemon/src/taker_cfd.rs (hunk 7 above)
Comment (rishflab) [reply daemon/src/taker_cfd.rs]: How would you know when to  trigger the rollover?

File: daemon/src/taker_cfd.rs (hunk 21 above)
Comment (rishflab) [reply daemon/src/taker_cfd.rs]: yep
